### PR TITLE
Tox and Bash improvements

### PR DESCRIPTION
**Pull request by XN137**
some LHF improvements to clean things up and make them stricter/fail early.

also partially reverts https://github.com/projectnessie/nessie-demos/pull/263 in cases where using the improved pip resolver is fine (it prevents dependabot from upgrading dependencies into unsupported combinations, see the comment about `Sphinx`).

We have to keep it for flink for now because the transitive dependencies there are a huge mess.

### CODE REVIEW

File: notebooks/requirements_dev.txt [hunk 1]
```diff
@@ -25,7 +25,7 @@ pytest==7.0.1
 pytest-mock==3.7.0
 pytest-mypy==0.9.1
 pytest-runner==6.0.0
-Sphinx==4.4.0
+Sphinx==4.3.2
```
Comment (XN137) [inline notebooks/requirements_dev.txt:28]: Sphinx 4.4.0 is not compatible with the latest version of flake8:
```
   The conflict is caused by:
      build 0.7.0 depends on importlib-metadata>=0.22; python_version < "3.8"
      flake8 4.0.1 depends on importlib-metadata<4.3; python_version < "3.8"
      pytest 7.0.1 depends on importlib-metadata>=0.12; python_version < "3.8"
      sphinx 4.4.0 depends on importlib-metadata>=4.4; python_version < "3.10"
```
using the proper / non-legacy pip resolver prevents dependabot from making invalid version upgrades in the future.

File: notebooks/tox.ini [hunk 2]
```diff
@@ -42,29 +41,23 @@ commands =
     pytest --basetemp={envtmpdir} -ra tests/test_nessie_iceberg_flink_demo_nba.py
 
 [testenv:hive]
-whitelist_externals=
-    /bin/bash
+allowlist_externals=bash
```
Comment (XN137) [inline notebooks/tox.ini:44]: this avoids warnings in github workflow

File: notebooks/tox.ini [hunk 3]
```diff
@@ -42,29 +41,23 @@ commands =
...
+passenv = TOXENV CI CODECOV_* JAVA_HOME
 deps =
-    --use-deprecated=legacy-resolver
     -r{toxinidir}/../docker/binder/requirements.txt
     -r{toxinidir}/requirements_dev.txt
 commands =
-    bash -ec 'chmod +x {toxinidir}/tests/scripts/start_hive'
-    bash -ec '{toxinidir}/tests/scripts/start_hive'
-    pip install -U pip
+    bash -ex {toxinidir}/tests/scripts/start_hive
```
Comment (XN137) [inline notebooks/tox.ini:52]: `chmod +x` is not needed when you call the script directly through bash

File: binder/Dockerfile [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 # Tag will be automatically generated through pre-commit hook if any changes
 # happened in the docker/ folder
-FROM ghcr.io/projectnessie/nessie-binder-demos:7fb962b09182c5f60ac2f80bbabb693290f3f741
+FROM ghcr.io/projectnessie/nessie-binder-demos:788b8cd170f0a67371e200dab75be3f97fd20d74
```
Comment (XN137) [inline binder/Dockerfile:5]: i ran the script to update this tag locally... but it wont built until this PR is merged an the workflow for `main` builds the latest images and publishes them to `ghcr`

**Review by omarsmak — APPROVED** [review]

**Comment by XN137** [comment]
i tested the docker logic locally like this:
```
cd ~/code/nessie-demos && bash .github/scripts/create_and_push_docker_image.sh projectnessie/nessie-binder-demos 982fe3adb1e8b58263c74d0b3fc4b6b9ab97f652 docker
```
building will succeed but publishing will fail (since i have no local credentials for the remote repo).

but then you can change `Dockerfile` to not fetch from the remote repo:
```
FROM projectnessie/nessie-binder-demos:982fe3adb1e8b58263c74d0b3fc4b6b9ab97f652
```
and thus you can run the notebooks locally with the locally built latest docker image:
```
jupyter-repo2docker .
```

this showed that i needed to pin the panads dependency to a different version for the flink environment but then i was able to run all 4 notebooks successfully.

it should be noted that running the notebooks via docker is not idempotent (cant run the same notebook twice) and also you cant run any other notebooks once one has run (due to leftover database/schema/nessie state).

@omarsmak please re-review the latest commits.

**Comment by omarsmak** [comment]
LGTM @XN137, thanks. Just one note, would it possible if you can document how to build the docker image and test the notebooks via Docker? I think that would be great to refer to everytime we need to test the changes locally.

**Comment by XN137** [comment]
> would it possible if you can document how to build the docker image and test the notebooks via Docker

can i do so in a follow-up PR?
the PR is widely spread as it is already imo :sweat_smile:

**Comment by omarsmak** [comment]
> > would it possible if you can document how to build the docker image and test the notebooks via Docker
> 
> can i do so in a follow-up PR? the PR is widely spread as it is already imo 😅

Sure no problems :)

**Review by omarsmak — DISMISSED** [review]

File: notebooks/requirements_dev.txt [hunk 5]
```diff
@@ -17,16 +17,13 @@
 assertpy==1.1
 bump2version==1.0.1
 build==0.7.0
-flake8==4.0.1
-ipython
+ipython==7.32.0
```
Comment (snazy) [inline notebooks/requirements_dev.txt:20]: Not against this change, but can you move this and the change in requirements_lint.txt to another PR?

File: docker/binder/requirements_flink.txt [hunk 6]
```diff
@@ -1,2 +1,4 @@
 -r requirements_base.txt
 apache-flink==1.12.1
+# flink requires pandas<1 see https://github.com/apache/flink/blob/release-1.12.1/flink-python/setup.py#L319-L320
```
Comment (snazy) [inline docker/binder/requirements_flink.txt:3]: Not against this change, but can you move this and the change in requirements_lint.txt to another PR?

**Review by snazy — COMMENTED** [review]
Good changes here.
Just would like to move the changes to the requirements.txt files into a separate PR.

**Comment by XN137** [comment]
@snazy ok i have removed the requirement changes for now and will move them to another PR

**Comment by XN137** [comment]
ok now the tests are failing because the current requirements are not compatible with the non-legacy pip resolver.

so i will split out this first:
https://github.com/projectnessie/nessie-demos/pull/281

then rebase this PR.

then do another PR with the remaining requirement changes hopefully :sweat_smile:

**Comment by XN137** [comment]
ok tests are passing now so this is ready for review again

**Review by omarsmak — APPROVED** [review]

**Pull request merged**